package/samplers/auto_sampler/_sampler.py

```diff
@@ -0,0 +1,230 @@
+            constraints is violated. A value equal to or smaller than 0 is considered feasible.
+            If ``constraints_func`` returns more than one value for a trial, that trial is
+            considered feasible if and only if all values are equal to 0 or smaller.
+
+            The ``constraints_func`` will be evaluated after each successful trial.
+            The function won't be called when trials fail or they are pruned, but this behavior is
+            subject to change in the future releases.
+    """
+
+    def __init__(
+        self,
+        *,
+        seed: int | None = None,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+    ) -> None:
+        self._rng = LazyRandomState(seed)
+        self._thread_local_sampler = _ThreadLocalSampler()
+        self._constraints_func = constraints_func
+
+    @property
+    def _sampler(self) -> BaseSampler:
+        if self._thread_local_sampler._sampler is None:
+            # NOTE(nabenabe): Do not do this process in the __init__ method because the
+            # substitution at the init does not update attributes in self._thread_local_sampler
+            # in each thread.
+            seed_for_random_sampler = self._rng.rng.randint(_MAXINT32)
+            self._sampler = RandomSampler(seed=seed_for_random_sampler)
+
+        return self._thread_local_sampler._sampler
+
+    @_sampler.setter
+    def _sampler(self, sampler: BaseSampler) -> None:
+        self._thread_local_sampler._sampler = sampler
+
+    def reseed_rng(self) -> None:
+        self._rng.rng.seed()
+        self._sampler.reseed_rng()
+
+    def _include_conditional_param(self, study: Study) -> bool:
+        trials = study._get_trials(
+            deepcopy=False, states=(TrialState.COMPLETE, TrialState.PRUNED), use_cache=True
+        )
+        if len(trials) == 0:
+            return False
+
+        param_key = set(trials[0].params)
+        return any(param_key != set(t.params) for t in trials)
+
+    def _determine_multi_objective_sampler(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> BaseSampler:
+        if isinstance(self._sampler, NSGAIISampler):
+            return self._sampler
+
+        seed = self._rng.rng.randint(_MAXINT32)
+        return NSGAIISampler(constraints_func=self._constraints_func, seed=seed)
+
+    def _determine_single_objective_sampler(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> BaseSampler:
+        if isinstance(self._sampler, TPESampler):
+            return self._sampler
+
+        seed = self._rng.rng.randint(_MAXINT32)
+        if (
+            self._constraints_func is not None
+            or any(isinstance(d, CategoricalDistribution) for d in search_space.values())
+            or self._include_conditional_param(study)
+        ):
+            # NOTE(nabenabe): The statement above is always true for Trial#1.
```

Comment:
Is this note correct?

Reply:
Simply forgot deleting the line, thank you for the catch!